apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import {
+  Colors,
+  JsonFile,
+  FileSystemStats,
+  IFileSystemCreateLinkOptions,
+  IColorableSequence
+} from '@rushstack/node-core-library';
+import * as crypto from 'crypto';
+import { Tslint, Typescript } from '@microsoft/rush-stack-compiler-3.7';
+import {
+  ExtendedTypescript,
+  IEmitResolver,
+  IExtendedEmitResult,
+  IEmitHost,
+  IEmitTransformers,
+  IExtendedProgram,
+  IExtendedSourceFile,
+  IResolveModuleNameResolutionHost
+} from './privateTypings/ExtendedTypescript';
+
+import { SubprocessRunnerBase } from '../../utilities/subprocess/SubprocessRunnerBase';
+import { Async } from '../../utilities/Async';
+import { resolvePackagePath } from '../../utilities/resolvePackage';
+import { IExtendedLinter } from './privateTypings/ExtendedTslint';
+import { IEmitModuleKindBase, ISharedTypescriptConfiguration } from '../../cli/actions/BuildAction';
+
+const ASYNC_LIMIT: number = 100;
+
+export interface ITypescriptBuilderConfiguration extends ISharedTypescriptConfiguration {
+  buildFolder: string;
+  typescriptToolPath: string;
+  tslintToolPath: string;
+  eslintToolPath: string;
+
+  lintingEnabled: boolean;
+
+  watchMode: boolean;
+
+  /**
+   * The path to the tsconfig file being built.
+   */
+  tsconfigPath: string;
+
+  /**
+   * The path to a tslint config file to use to lint the files being compiled, if one exists
+   */
+  tslintPath: string | undefined;
+
+  /**
+   * The path of project's build cache folder
+   */
+  buildCacheFolder: string;
+}
+
+interface ICachedEmitModuleKind<TModuleKind> extends IEmitModuleKindBase<TModuleKind> {
+  cacheOutFolderPath: string;
```

Comment:
The purpose of these interfaces is not obvious.  It would be good to document them.
For each setting:
- What does it do?
- If it's a string, show an example value
- If it's a path, what is it relative to?
- If it's `undefined` then what is the behavior?

Reply:
The TypescriptBuilder.ts is likely to be a primary focus for optimizations, so it would be worthwhile to make the code as easy for people to understand as possible.